### PR TITLE
fix: inplace update failed due to virtual kubelet in PUB

### LINKED ISSUE #1262: [feature request] 在virtual-kubelet节点中，优化原地升级支持
**在InPlaceUpdateEnvFromMetadata特性打开的情况下，由于virtual-kubelet节点不支持安装kruise daemon，会导致原地更新卡住，希望可以优化该场景**:
场景：在virtual-kubelet节点上打开InPlaceUpdateEnvFromMetadata特性，修改镜像版本，pod会进入原地升级，InPlaceUpdateReady condition会从true变为false，但在Container重启Ready后，InPlaceUpdateReady condition不会恢复成true。
原因分析：通过追踪源码和数据，发现是进了下图中的代码分支：
image
pod上的annotation如下：
image
这里因为将InPlaceUpdateEnvFromMetadata打开，所以updateEnvFromMetadata为true，但又没有kruise daemon来通过annotation来上报状态，所以这里的runtimeContainerMetaSet 为空，导致原地升级流程卡住。

### PR DESCRIPTION
**Pull request by chengjoey**
### Ⅰ. Describe what this PR does
 fix inplace update failed due to inplace update on vk node
skip inPlaceUpdateState check when pod is running on vk node

### Ⅱ. Does this pull request fix one issue?
fix #1262

### Ⅲ. Describe how to verify it
inplace  update on vk node

### CODE REVIEW

File: pkg/control/pubcontrol/pub_control_utils.go [hunk 1]
```diff
@@ -43,6 +44,21 @@ const (
 	MaxUnavailablePodSize = 2000
 )
 
+var (
```
Comment (hantmac) [inline pkg/control/pubcontrol/pub_control_utils.go:47]: How about making `var` together?
```
var (
labelNodeTypeKey = "type"
virtualNodeType = "virtual-kubelet"

taintVirtualToFind = &corev1.Taint{
		Key:    "virtual-kubelet.io/provider",
		Effect: corev1.TaintEffectNoSchedule,
	}
)
```

File: pkg/control/pubcontrol/pub_control_utils.go [hunk 2]
```diff
@@ -175,6 +191,19 @@ func PodUnavailableBudgetValidatePod(client client.Client, control PubControl, p
...
+func isVirtualKubeletNode(cli client.Client, nodeName string) bool {
+	if len(nodeName) == 0 {
+		return false
+	}
+	node := &corev1.Node{}
+	if err := cli.Get(context.Background(), client.ObjectKey{Name: nodeName}, node); err != nil {
+		klog.Errorf("get node(%s) failed: %s", nodeName, err.Error())
+		return false
+	}
+	return node.Labels[labelNodeTypeKey] == virtualNodeType && taints.TaintExists(node.Spec.Taints, taintVirtualToFind)
```
Comment (hantmac) [inline pkg/control/pubcontrol/pub_control_utils.go:204]: I think there is no need to check the taints on node, ` virtualNode` label is enough?

File: pkg/control/pubcontrol/pub_control_utils.go (hunk 1 above)
Comment (chengjoey) [reply pkg/control/pubcontrol/pub_control_utils.go:46]: done

File: pkg/control/pubcontrol/pub_control_utils.go (hunk 2 above)
Comment (chengjoey) [reply pkg/control/pubcontrol/pub_control_utils.go]: done, removed the judgment of taint， `virtualNode` label in enough

**Review by hantmac — APPROVED** [review]
/LGTM

**Comment by zmberg** [comment]
This situation is still a bit more complicated, you can't skip the vk node so easily, many of our internal scenarios vk can be upgraded in place

**Comment by chengjoey** [comment]
> This situation is still a bit more complicated, you can't skip the vk node so easily, many of our internal scenarios vk can be upgraded in place

Thanks! @zmberg, I added a judgment, only the pod running on the virtual node and `UpdateEnvFromMetadata` is turned on will skip the check

**Pull request closed without merging**